### ISSUE TITLE
Allow mutating `StivaleSmpInfo`

### ISSUE THREAD

**Issue by JCapucho** [Andy-Python-Programmer/stivale#4]
The `StivaleSmpInfo` can be written to allow for example setting the AP stack and instruction pointer. the struct itself is writable but the only way to access it trough `StivaleSmpTag` which doesn't provide any API to get a mutable reference (one could possibly build a mutable slice directly from `smp_info_array`). It would also be needed for `StivaleStruct` to return a  `StivaleSmpTag` that allowed mutation trough `smp` (It's possible using `get_tag` and some casts but it would be nicer to have an API for it).

I would be available to implement just wanted to get your opinion on how the API should look.

**Comment by Andy-Python-Programmer** [comment]
> The `StivaleSmpInfo` can be written to allow for example setting the AP stack and instruction pointer. the struct itself is writable but the only way to access it trough `StivaleSmpTag` which doesn't provide any API to get a mutable reference (one could possibly build a mutable slice directly from `smp_info_array`). It would also be needed for `StivaleStruct` to return a  `StivaleSmpTag` that allowed mutation trough `smp`

Alrighty, I think making `StivaleStruct` returning a mutable reference to the struct would be fine but will have more safety concerns (for example someone changing the some field where you believe you have the copy of the value but you are actually mutating the value in the struct itself). I would say making it return a mutable slice from a function called `smp_info_array_mut` would have much lesser safety concerns (though still will have some). In that case add a `SAFTEY` doc on top of the `smp_info_array_mut` function that:

- The kernel has not reclaimed `BOOTLOADER_RECLAIMABLE` memory (since the boot info structures allocated are marked as bootloader reclaimable).
- They actually know what they are doing. For example if they are mutating the AP stack, the address set to is correct.

Or another option would be to add manual setter functions for setting the values.

> It's possible using `get_tag` and some casts but it would be nicer to have an API for it

Can you expand a bit more on this?

> I would be available to implement just wanted to get your opinion on how the API should look.

Sure!

**Comment by JCapucho** [comment]
I've opened #5 as a draft of what the API would look like.

> The kernel has not reclaimed BOOTLOADER_RECLAIMABLE memory (since the boot info structures allocated are marked as bootloader reclaimable).

FIY this is the responsibility of whoever is reclaiming the memory (no references to the reclaimed memory must exist after it being reclaimed)

> Can you expand a bit more on this?

Sure it's how the PR implements it

```rust
pub fn smp_mut(&mut self) -> Option<&'static mut StivaleSmpTag> {
    self.get_tag(0x34d1d96339647025)
        .map(|addr| unsafe { &mut *(addr as *mut StivaleSmpTag) })
}
```

Doable but with a proper API it looks better IMO

**Comment by Andy-Python-Programmer** [comment]
Published the fix in `0.2.6`!
